[2.9] Change rancher base image to bci-micro

pjbgf opened the pull request:
The previous base image (`bci-base`) contains several packages that are not required for the `rancher` container image.

The changes aim to decrease the long-term amount of base image CVEs that increases as a given container image version ages.

Follow-up from https://github.com/rancher/rancher/pull/44044.

UPDATE: It also reintroduces `umount`, `mount` and `unshare` to the `rancher/agent` image.

pjbgf committed:
Change base image to bci-micro

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

pjbgf committed:
Remove use of --gpg-auto-import-keys

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

macedogm commented:
Note: will review this week, then we can get more folks to review this too.

pjbgf commented:
Rebased the PR.

Oats87 reviewed (dismissed):
high level LGTM, I do expect that users/support may find a difference in attempting to debug their environments so we'll definitely want to make sure we call this out.
I'm not sure if we have an actual dependency list of what would be required for Rancher to function (binary wise) within the image, so we'll need to make sure this is thoroughly tested by QA before release IMO

pjbgf commented:
@Oats87 +1 on the testing front. For reference, here are the binaries that would no longer be shipped with `rancher/rancher`:

```
$ container-diff diff --type=file remote://rancher/rancher:v2.9-head daemon://rancher/rancher:87db95220 | grep "/bin/"
/bin/cpio                                   13B
/bin/dmesg                                  14B
/bin/kill                                   13B
/bin/login                                  42.2K
/bin/more                                   13B
/bin/mount                                  14B
/bin/pgrep                                  26.7K
/bin/pkill                                  26.7K
/bin/ps                                     131.2K
/bin/rpm                                    12B
/bin/su                                     11B
/bin/umount                                 15B
/usr/bin/appdata2solv                       285.2K
/usr/bin/cal                                62.6K
/usr/bin/chkconfig                          19.9K
/usr/bin/chmem                              62.2K
/usr/bin/choom                              50.2K
/usr/bin/chrt                               34.2K
/usr/bin/clear                              10.3K
/usr/bin/col                                30.2K
/usr/bin/colcrt                             14.2K
/usr/bin/colrm                              26.2K
/usr/bin/column                             50.2K
/usr/bin/comps2solv                         281.3K
/usr/bin/container-suseconnect              7.7M
/usr/bin/cpio                               154.8K
/usr/bin/deltainfoxml2solv                  280.9K
/usr/bin/dmesg                              78.5K
/usr/bin/dumpsolv                           252.6K
/usr/bin/eject                              78.2K
/usr/bin/fallocate                          30.2K
/usr/bin/filesize                           68B
/usr/bin/fincore                            30.3K
/usr/bin/flock                              30.3K
/usr/bin/free                               18.4K
/usr/bin/g13                                190K
/usr/bin/gendiff                            499B
/usr/bin/get_kernel_version                 10.2K
/usr/bin/getopt                             30.2K
/usr/bin/gpg                                4B
/usr/bin/gpg-agent                          399.7K
/usr/bin/gpg-connect-agent                  144.5K
/usr/bin/gpg-wks-server                     188.6K
/usr/bin/gpg-zip                            3.4K
/usr/bin/gpg2                               1M
/usr/bin/gpgconf                            160K
/usr/bin/gpgparsemail                       26.6K
/usr/bin/gpgscm                             211.9K
/usr/bin/gpgsm                              498.3K
/usr/bin/gpgsplit                           71.6K
/usr/bin/gpgtar                             136.8K
/usr/bin/gpgv                               5B
/usr/bin/gpgv2                              429.1K
/usr/bin/hardlink                           38.3K
/usr/bin/hexdump                            54.3K
/usr/bin/i386                               7B
/usr/bin/import-suse-build-key              1K
/usr/bin/infocmp                            62.7K
/usr/bin/installation_sources               754B
/usr/bin/installcheck                       658.9K
/usr/bin/ionice                             26.2K
/usr/bin/ipcmk                              30.3K
/usr/bin/ipcrm                              30.2K
/usr/bin/ipcs                               70.2K
/usr/bin/irqtop                             46.4K
/usr/bin/isosize                            26.2K
/usr/bin/kbxutil                            156.2K
/usr/bin/kill                               38.2K
/usr/bin/last                               46.2K
/usr/bin/lastb                              4B
/usr/bin/line                               10.2K
/usr/bin/linux32                            7B
/usr/bin/linux64                            7B
/usr/bin/look                               14.2K
/usr/bin/lscpu                              118.2K
/usr/bin/lsipc                              90.2K
/usr/bin/lsirq                              34.4K
/usr/bin/lslocks                            38.6K
/usr/bin/lsmem                              62.2K
/usr/bin/lsns                               46.2K
/usr/bin/mcookie                            30.3K
/usr/bin/mergesolv                          300.7K
/usr/bin/mesg                               14.2K
/usr/bin/mkinfodir                          7.7K
/usr/bin/more                               54.2K
/usr/bin/mount                              54.2K
/usr/bin/mountpoint                         14.2K
/usr/bin/namei                              30.2K
/usr/bin/nsenter                            34.5K
/usr/bin/old                                1.2K
/usr/bin/pgrep                              10B
/usr/bin/pinentry                           3K
/usr/bin/pinentry-curses                    63.6K
/usr/bin/pinentry-tty                       55.3K
/usr/bin/pkill                              10B
/usr/bin/pmap                               30.6K
/usr/bin/prlimit                            34.8K
/usr/bin/ps                                 7B
/usr/bin/pwdx                               10.4K
/usr/bin/rename                             18.2K
/usr/bin/renice                             14.2K
/usr/bin/repo2solv                          441K
/usr/bin/repo2solv.sh                       9B
/usr/bin/repomdxml2solv                     281.2K
/usr/bin/reset                              4B
/usr/bin/rev                                14.2K
/usr/bin/rpm                                18.9K
/usr/bin/rpm2cpio                           10.4K
/usr/bin/rpmdb                              15.1K
/usr/bin/rpmdb2solv                         398.6K
/usr/bin/rpmgraph                           14.8K
/usr/bin/rpmkeys                            14.9K
/usr/bin/rpmlocate                          336B
/usr/bin/rpmmd2solv                         314.6K
/usr/bin/rpmqpack                           10.4K
/usr/bin/rpmquery                           3B
/usr/bin/rpms2solv                          393.5K
/usr/bin/rpmsign                            15.1K
/usr/bin/rpmspec                            15.1K
/usr/bin/rpmverify                          3B
/usr/bin/safe-rm                            901B
/usr/bin/safe-rmdir                         7B
/usr/bin/scdaemon                           426.7K
/usr/bin/script                             66.2K
/usr/bin/scriptlive                         54.2K
/usr/bin/scriptreplay                       38.3K
/usr/bin/setarch                            18.5K
/usr/bin/setpriv                            50.2K
/usr/bin/setsid                             14.2K
/usr/bin/setterm                            42.2K
/usr/bin/skill                              26.6K
/usr/bin/slabtop                            18.5K
/usr/bin/snice                              26.6K
/usr/bin/ssh.hmac                           32B
/usr/bin/su                                 70.2K
/usr/bin/susetags2solv                      357.2K
/usr/bin/tabs                               14.4K
/usr/bin/taskset                            34.2K
/usr/bin/testsolv                           737.4K
/usr/bin/tload                              14.5K
/usr/bin/toe                                14.4K
/usr/bin/top                                122.9K
/usr/bin/tput                               18.5K
/usr/bin/tset                               22.5K
/usr/bin/tzselect                           21.8K
/usr/bin/uclampset                          58.2K
/usr/bin/ul                                 18.2K
/usr/bin/umount                             34.2K
/usr/bin/uname26                            7B
/usr/bin/unshare                            46.5K
/usr/bin/updateinfoxml2solv                 285.3K
/usr/bin/utmpdump                           26.2K
/usr/bin/uuidgen                            14.2K
/usr/bin/uuidparse                          38.2K
/usr/bin/vmstat                             34.5K
/usr/bin/w                                  18.5K
/usr/bin/wall                               34.2K
/usr/bin/watch                              27.2K
/usr/bin/watchgnupg                         14.5K
/usr/bin/wdctl                              62.3K
/usr/bin/whereis                            30.7K
/usr/bin/write                              18.2K
/usr/bin/x86_64                             7B
/usr/bin/yzpper                             6B
/usr/bin/zypp-CheckAccessDeleted            22.6K
/usr/bin/zypp-NameReqPrv                    99.9K
/usr/bin/zypper                             3.2M
/usr/lib/restricted/bin/hostname            13B
```

moio commented:
I am not qualified to answer if this use of pkill could be problematic or not:

https://github.com/rancher/rancher/blob/a6da228939daa9dd1444d36190fcb9a0338a2ea4/pkg/controllers/dashboard/helm/operations.go#L136

moio commented:
Also this use of `umount`:

https://github.com/rancher/rancher/blob/a6da228939daa9dd1444d36190fcb9a0338a2ea4/pkg/agent/clean/node_notwindows.go#L222

moio commented:
Also `unshare` here:

https://github.com/rancher/rancher/blob/a6da228939daa9dd1444d36190fcb9a0338a2ea4/package/kubectl-shell.sh#L24

pjbgf commented:
> I am not qualified to answer if this use of pkill could be problematic or not:
> 
> https://github.com/rancher/rancher/blob/a6da228939daa9dd1444d36190fcb9a0338a2ea4/pkg/controllers/dashboard/helm/operations.go#L136

This is being executed against the `proxy` pod, not Rancher's:
https://github.com/rancher/rancher/blob/a6da228939daa9dd1444d36190fcb9a0338a2ea4/pkg/controllers/dashboard/helm/operations.go#L134-L136

pjbgf committed:
agent: Reintroduce unshare, mount and umount
When the rancher agent base image was changed to bci-micro (fe2c907021),
several binaries were removed from the final image due to the difference
between bci-base and bci-micro.

This change reintroduces unshare, mount and umount, which are needed to
support agent clean-up the agent shell feature.

xref:
- https://github.com/rancher/rancher/commit/a350fe70ddfde9f747d169ce9b5c80c540dd3400
- https://github.com/rancher/rancher/commit/5eb2d4e8a3f17bd15f83a45d6b9a161055b89222
- https://github.com/rancher/rancher/commit/52da4437d2c6f7d299c17f0f7f5b241160011cb3

Signed-off-by: Paulo Gomes <paulo.gomes@suse.com>

pjbgf commented:
@moio nice spot! thank you for taking a look at this PR and pointing out the cases above.

As mentioned before, the `pkill` is being executed against the `proxy` container which uses the `shell` image, so it should not be affected by this change.

Looking deeper at `umount` and `unshare` (plus `mount`), I noticed that they are not needed for `rancher/rancher`, but they are for `rancher/agent`. I added an additional commit to re-introduce them with a commit message that hopefully explains where they are used.

PTAL

moio reviewed (approved):
FWIW I tried grepping the Rancher codebase for (non obvious) names of removed executables and those seem covered now.

I guess there is no rock-solid way to ensure the absence of regressions without trying, so I am approving this now.

pjbgf commented:
CI failure is orthogonal as per https://github.com/rancher/rancher/issues/45159:
```
--- FAIL: TestUIPluginSuite (39.43s)
    --- PASS: TestUIPluginSuite/TestGetIndexAuthenticated (0.00s)
    --- PASS: TestUIPluginSuite/TestGetIndexUnauthenticated (0.00s)
    --- PASS: TestUIPluginSuite/TestGetSingleExtensionAuthenticated (0.00s)
    --- FAIL: TestUIPluginSuite/TestGetSingleExtensionUnauthenticated (0.00s)
FAIL
FAIL	github.com/rancher/rancher/tests/v2/integration/catalogv2	412.797s
```

pjbgf commented:
Merging as per off-line chat with @macedogm, and the approvals above.
